File: cours/S7/Algo proba/Rand_Algo/QS.py
from numpy import *;
from random import *;


def echange(T, i, j):
    k = T[i];
    T[i] = T[j];
    T[j] = k;



def Partition( T, a, b, k):
    p = T[k];
    echange(T, k, b);
    start = a;
    for i in range(a, b):
        if T[i] < p:
            echange(T, i, start);
            start += 1;
    echange(T, start, b);
    return start;



def QuickSort( T, a, b):
    if a<b:
        k = Partition(T, a, b, a);
        QuickSort(T, a, k-1);
        QuickSort(T, k+1, b);
    

def QuickSortRandom( T, a, b):
    if a<b:
        k = Partition(T, a, b, randint(a,b));
        QuickSort(T, a, k-1);
        QuickSort(T, k+1, b);
    

def ProdArrayRandom( n):
    T = array([randint(0,n)]);
    
    for i in range(1,n):
        T = append(T, array([randint(0,n)]));
    return T;





#test

T = array([1,3,2,7,8,98,76,33]);

print T;

QuickSortRandom( T, 0, 7);

print T;

print ProdArrayRandom(15);

File: cours/S7/Algo proba/Rand_Algo/QuickSort/C/QS.c
#include <stdio.h>
#include <stdlib.h>


int rand_int(int a, int b)
{

  return rand()%(b-a) + a;

}




void echange(int T[], int i, int j)
{
  int k = T[j];
  T[j] = T[i];
  T[i] = k;
}


int Partition(int T[], int a, int b, int k)
{
  int p = T[k];
  echange(T, k, b);
  int start = a;
  for(int i = a; i<b; i++)
    {
      if (T[i]<p)
	{
	  echange(T,i,start);
	  start++;
	}
    }
  echange(T, start, b);
  return start;
}



void QuickSort(int T[], int a, int b)
{
  if(a<b)
    {
      int k = Partition(T, a, b, a);
      QuickSort(T,a,k-1);
      QuickSort(T,k+1,b);
    }
	
}


void RandomQuickSort(int T[], int a, int b)
{
  if(a<b)
    {
      int p = rand_int(a,b);
      int k = Partition(T, a, b, p);
      QuickSort(T,a,k-1);
      QuickSort(T,k+1,b);
    }
}

int main()
{
  int T[12] = {3, 4, 1, 4, 5, 6, 7, 8, 76, 9, 54, 13};
  
  //QuickSort(T,0,11);
  RandomQuickSort(T,0,11);

 
  return 0;
}
